src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -2051,14 +2051,15 @@ internal override bool FilterAndAppendAndFreeDiagnostics(DiagnosticBag accumulat
         /// Filter out warnings based on the compiler options (/nowarn, /warn and /warnaserror) and the pragma warning directives.
         /// </summary>
         /// <returns>True when there is no error.</returns>
-        private bool FilterAndAppendDiagnostics(DiagnosticBag accumulator, IEnumerable<Diagnostic> incoming)
+        private bool FilterAndAppendDiagnostics(DiagnosticBag accumulator, IEnumerable<Diagnostic> incoming, bool includeDiagnosticsWithSourceSuppression = false)
         {
             bool hasError = false;
 
             foreach (Diagnostic d in incoming)
             {
                 var filtered = _options.FilterDiagnostic(d);
-                if (filtered == null)
+                if (filtered == null ||
+                    (!includeDiagnosticsWithSourceSuppression && filtered.HasSourceSuppression))
```

Comment:
I think the special casing feels a bit strange from the internal API / implementation perspective. Diagnostics suppressed via nowarn or ruleset will have ReportDiagnostic.Suppressed whereas those suppressed in source will have ReportDiagnostic.Default and we have to always check an additional flag to determine if it was suppressed in source... This is probably fine for now - but it would be nice if we can rationalize this somehow to avoid mistakes in the future.